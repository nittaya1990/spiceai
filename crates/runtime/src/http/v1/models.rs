/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
use std::{collections::HashMap, sync::Arc};

use app::App;
use axum::{
    extract::Query,
    http::status,
    response::{IntoResponse, Json, Response},
    Extension,
};
use csv::Writer;
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

use crate::{status::ComponentStatus, Runtime};

use super::Format;

#[derive(Debug, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::IntoParams))]
pub struct ModelsQueryParams {
    /// The format of the response (e.g., `json` or `csv`).
    #[serde(default)]
    pub format: Format,

    /// If true, includes the status of each model in the response.
    #[serde(default)]
    pub status: bool,
}

#[derive(Debug, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[serde(rename_all = "snake_case")]
pub(crate) struct OpenAIModelResponse {
    object: String,
    data: Vec<OpenAIModel>,
}

#[derive(Debug, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[serde(rename_all = "snake_case")]
pub(crate) struct OpenAIModel {
    /// The name of the model
    id: String,

    /// The type of the model (always `model`)
    object: String,

    /// The source from which the model was loaded (e.g., `openai`, `spiceai`)
    owned_by: String,

    /// The datasets associated with this model, if any
    datasets: Option<Vec<String>>,

    /// The status of the model (e.g., `ready`, `initializing`, `error`)
    status: Option<ComponentStatus>,
}

/// List Models
///
/// List all models, both machine learning and language models, available in the runtime.
#[cfg_attr(feature = "openapi", utoipa::path(
    get,
    path = "/v1/models",
    operation_id = "get_models",
    tag = "AI",
    params(ModelsQueryParams),
    responses(
        (status = 200, description = "List of models in JSON format", content((
            OpenAIModelResponse = "application/json",
            example = json!({
                "object": "list",
                "data": [
                    {
                        "id": "gpt-4",
                        "object": "model",
                        "owned_by": "openai",
                        "datasets": null,
                        "status": "ready"
                    },
                    {
                        "id": "text-embedding-ada-002",
                        "object": "model",
                        "owned_by": "openai-internal",
                        "datasets": ["text-dataset-1", "text-dataset-2"],
                        "status": "ready"
                    }
                ]
            })
        ), (
            String = "text/csv",
            example = "
id,object,owned_by,datasets,status
gpt-4,model,openai,,ready
text-embedding-ada-002,model,openai-internal,\"text-dataset-1,text-dataset-2\",ready
"
        ))),
        (status = 500, description = "Internal server error occurred while processing models", content((
            serde_json::Value = "application/json",
            example = json!({
                "error": "App not initialized"
            })
        )))
    )
))]
pub(crate) async fn get(
    Extension(app): Extension<Arc<RwLock<Option<Arc<App>>>>>,
    Extension(rt): Extension<Arc<Runtime>>,
    Query(params): Query<ModelsQueryParams>,
) -> Response {
    let statuses = if params.status {
        rt.status.get_model_statuses()
    } else {
        HashMap::default()
    };
    let models = match app.read().await.as_ref() {
        Some(a) => a
            .models
            .iter()
            .map(|m| {
                let d = if m.datasets.is_empty() {
                    None
                } else {
                    Some(m.datasets.clone())
                };
                OpenAIModel {
                    id: m.name.clone(),
                    object: "model".to_string(),
                    owned_by: m.from.clone(),
                    datasets: d,
                    status: statuses.get(&m.name).copied(),
                }
            })
            .collect::<Vec<OpenAIModel>>(),
        None => {
            return (
                status::StatusCode::INTERNAL_SERVER_ERROR,
                "App not initialized",
            )
                .into_response();
        }
    };

    match params.format {
        Format::Json => (
            status::StatusCode::OK,
            Json(OpenAIModelResponse {
                object: "list".to_string(),
                data: models,
            }),
        )
            .into_response(),
        Format::Csv => match convert_details_to_csv(&models) {
            Ok(csv) => (status::StatusCode::OK, csv).into_response(),
            Err(e) => {
                tracing::error!("Error converting to CSV: {e}");
                (status::StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response()
            }
        },
    }
}

fn convert_details_to_csv(models: &[OpenAIModel]) -> Result<String, Box<dyn std::error::Error>> {
    let mut w = Writer::from_writer(vec![]);
    for d in models {
        let _ = w.serialize(d);
    }
    w.flush()?;
    Ok(String::from_utf8(w.into_inner()?)?)
}
