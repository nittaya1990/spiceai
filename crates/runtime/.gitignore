abc-sqlite.db-wal
abc-sqlite.db-shm
