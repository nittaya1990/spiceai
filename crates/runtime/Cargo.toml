[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow.workspace = true
arrow-csv = "53.3.0"
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow-schema = { workspace = true, features = ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-dynamodb = { version = "1", optional = true }
aws-sdk-secretsmanager = { version = "1", optional = true }
aws-sdk-sts = { version = "1", optional = true }
axum.workspace = true
axum-extra = { version = "0.9.6", features = ["typed-header"] }
base64.workspace = true
bytes.workspace = true
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "6.1.0"
data_components = { path = "../data_components" }
datafusion.workspace = true
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
    "bundled",
    "r2d2",
    "vtab",
    "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
governor = "0.8.0"
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http.workspace = true
http-body-util = "0.1.2"
hyper = "1.6.0"
hyper-util = { version = "0.1.6", features = ["service"] }
iceberg.workspace = true
iceberg-catalog-rest.workspace = true
imap = { workspace = true, optional = true }
indexmap.workspace = true
itertools.workspace = true
jsonwebtoken = "9.3.0"
keyring = { version = "3.6.1", features = [
    "apple-native",
    "windows-native",
    "linux-native",
], optional = true }
llms = { path = "../llms" }
logos = "0.15.0"
mcp-client = { git = "https://github.com/block/goose", package = "mcp-client", optional = true }
mcp-core = { git = "https://github.com/block/goose", package = "mcp-core", optional = true }
mediatype = "0.19.18"
model_components = { path = "../model_components" }
mysql_async = { workspace = true, optional = true }
notify = "7.0.0"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry-prometheus.workspace = true
opentelemetry-proto = { version = "0.27", features = [
    "gen-tonic-messages",
    "gen-tonic",
    "metrics",
] }
opentelemetry_sdk.workspace = true
otel-arrow = { path = "../otel-arrow" }
package = { path = "../package" }
percent-encoding.workspace = true
pin-project.workspace = true
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
    "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
runtime-auth = { path = "../runtime-auth" }
rusqlite = { workspace = true, optional = true }
rustls.workspace = true
rustls-pemfile.workspace = true
schemars = "0.8.19"
secrecy.workspace = true
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tiberius = { workspace = true, optional = true }
tokio.workspace = true
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio-stream.workspace = true
tokio-util = { workspace = true, optional = true }
tonic.workspace = true
tonic-health.workspace = true
tower.workspace = true
tower-http.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.4"
util = { path = "../util" }
utoipa = { version = "5", features = ["axum_extras"], optional = true }
utoipa-swagger-ui = { version = "8", features = ["axum"], optional = true }

uuid = { workspace = true, features = ["v7"] }
x509-certificate.workspace = true

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "=7.0.13"
azure_core = "0.21.0"
azure_storage = "0.21.0"
azure_storage_blobs = "0.21.0"
bollard = "0.18.1"
dirs = "5.0.1"
flightrepl = { path = "../flightrepl" }
hf-hub = { workspace = true }
insta.workspace = true
jsonpath-rust = "0.7.3"
opentelemetry_sdk = { workspace = true, features = ["rt-tokio-current-thread"] }
paste.workspace = true
scopeguard = "1.2.0"
serde_json = { workspace = true, features = ["preserve_order"] }
spice-cloud = { path = "../spice_cloud" }
test-framework = { path = "../test-framework" }
tokio = { workspace = true, features = ["time", "test-util"] }
tracing-opentelemetry.workspace = true
tracing-subscriber.workspace = true

[features]
aws-secrets-manager = [
    "dep:aws-config",
    "dep:aws-sdk-secretsmanager",
    "dep:aws-sdk-sts",
]
clickhouse = [
    "dep:clickhouse-rs",
    "db_connection_pool/clickhouse",
    "arrow_sql_gen/clickhouse",
    "data_components/clickhouse",
]
cuda = ["llms/cuda"]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = ["openapi", "otel-arrow/dev", "mcp"]
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
dynamodb = ["dep:aws-sdk-dynamodb", "data_components/dynamodb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
imap = ["data_components/imap", "dep:imap"]
keyring-secret-store = ["dep:keyring"]
mcp = ["dep:mcp-client", "dep:mcp-core"]
metal = ["llms/metal"]
models = ["model_components/full"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
openapi = ["dep:utoipa-swagger-ui", "dep:utoipa"]
postgres = [
    "db_connection_pool/postgres",
    "data_components/postgres",
    "datafusion-table-providers/postgres",
]
postgres-write = []
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
    "dep:snowflake-api",
    "db_connection_pool/snowflake",
    "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
    "dep:tokio-rusqlite",
    "db_connection_pool/sqlite",
    "data_components/sqlite",
    "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"

[[bench]]
harness = false
name = "vector_search"

[target.'cfg(windows)'.dependencies]
winver = "1.0.0"
