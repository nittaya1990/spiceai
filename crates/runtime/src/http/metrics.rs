/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::sync::LazyLock;

use opentelemetry::{
    global,
    metrics::{Counter, Histogram, Meter},
};

static METER: LazyLock<Meter> = LazyLock::new(|| global::meter("http"));

/// Deprecated, to be removed in the future
pub(crate) static REQUESTS_TOTAL: LazyLock<Counter<u64>> = LazyLock::new(|| {
    METER
        .u64_counter("http_requests_total")
        .with_description("Number of HTTP requests. Deprecated, use http_requests instead.")
        .build()
});

pub(crate) static REQUESTS: LazyLock<Counter<u64>> = LazyLock::new(|| {
    METER
        .u64_counter("http_requests")
        .with_description("Number of HTTP requests.")
        .build()
});

pub(crate) static REQUESTS_DURATION_MS: LazyLock<Histogram<f64>> = LazyLock::new(|| {
    METER
        .f64_histogram("http_requests_duration_ms")
        .with_unit("ms")
        .build()
});
