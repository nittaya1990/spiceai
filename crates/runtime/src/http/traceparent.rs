/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use axum::http::HeaderMap;
use opentelemetry::trace::{SpanId, TraceId};
use tracing::Span;

const MAX_VERSION: u8 = 254;
const TRACEPARENT_HEADER: &str = "traceparent";

/// Use the span context from the traceparent header to override the `trace_id` & `parent_span_id` columns in the task history table.
///
/// Errors are not returned, invalid traceparent headers must be ignored.
///
/// This should not be used for any span within a HTTP API that has [HTTP Spans](https://opentelemetry.io/docs/specs/semconv/http/http-spans/) created, as they are incompatible (both the `span` input and the span created for the HTTP handler will have the same `parent_span_id`, even though the `input` span would become a child of the HTTP span).
pub(super) fn override_task_history_with_traceparent(span: &Span, headers: &HeaderMap) {
    match extract_traceparent(headers) {
        Ok(Some((trace_id, span_id))) => {
            tracing::info!(target: "task_history", parent: span, trace_id = %trace_id, parent_id = %span_id);
        }
        Err(e) => {
            tracing::warn!("Recieved invalid `traceparent` HTTP header: {e}");
        }
        _ => {}
    }
}

fn extract_traceparent(headers: &HeaderMap) -> Result<Option<(TraceId, SpanId)>, String> {
    let Some(header_value) = headers.get(TRACEPARENT_HEADER).map(|v| v.to_str()) else {
        return Ok(None);
    };
    let header_value = header_value.map_err(|e| {
        format!("In traceparent header, invalid traceparent header value, expected string, got {e}")
    })?;
    let parts = header_value.split_terminator('-').collect::<Vec<&str>>();
    // Ensure parts are not out of range.
    if parts.len() < 4 {
        return Err(format!(
            "In traceparent header, invalid traceparent header, expected 4 parts, got {}",
            parts.len()
        ));
    }

    // Ensure version is within range, for version 0 there must be 4 parts.
    let version = u8::from_str_radix(parts[0], 16).map_err(|e| {
        format!("In traceparent header, invalid traceparent version, expected hex value, got {e}")
    })?;
    if version > MAX_VERSION || version == 0 && parts.len() != 4 {
        return Err(format!(
            "In traceparent header, invalid traceparent version {version}"
        ));
    }

    // Ensure trace id is lowercase
    if parts[1].chars().any(|c| c.is_ascii_uppercase()) {
        return Err(format!(
            "In traceparent header, invalid trace id. Expected lowercase hex value, got {}",
            parts[1]
        ));
    }

    // Parse trace id section
    let trace_id = TraceId::from_hex(parts[1]).map_err(|e| {
        format!("In traceparent header, invalid trace id. Expected 32 character hex value: {e}")
    })?;

    // Ensure span id is lowercase
    if parts[2].chars().any(|c| c.is_ascii_uppercase()) {
        return Err(format!(
            "In traceparent header, invalid span id. Expected lowercase hex value, got {}",
            parts[2]
        ));
    }

    // Parse span id section
    let span_id = SpanId::from_hex(parts[2]).map_err(|e| {
        format!("In traceparent header, invalid span id. Expected 16 character hex value: {e}")
    })?;

    // Parse trace flags section solely to ensure they're valid.
    let _ = u8::from_str_radix(parts[3], 16).map_err(|e| {
        format!("In traceparent header, invalid trace flags. Expected hex value, got {e}")
    })?;

    Ok(Some((trace_id, span_id)))
}
