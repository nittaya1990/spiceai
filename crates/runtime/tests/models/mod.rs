/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use arrow::{array::StringArray, util::pretty::pretty_format_batches};
use async_openai::types::EmbeddingInput;
use futures::TryStreamExt;
use rand::Rng;
use reqwest::{header::HeaderMap, Client};
use runtime::{config::Config, get_params_with_secrets, Runtime};
use secrecy::SecretString;
use snafu::ResultExt;
use spicepod::component::{
    dataset::{acceleration::Acceleration, Dataset},
    params::Params,
};
use std::sync::Arc;
use std::{
    collections::HashMap,
    net::{IpAddr, Ipv4Addr, SocketAddr},
};

use serde_json::{json, Value};
mod embedding;
mod hf;
mod local;
mod openai;

mod nsql {
    use chrono::{DateTime, Utc};
    use http::{
        header::{ACCEPT, CONTENT_TYPE},
        HeaderMap, HeaderValue,
    };
    use opentelemetry_sdk::trace::TracerProvider;

    use crate::models::http_post;

    pub struct TestCase {
        pub name: &'static str,
        pub body: serde_json::Value,
    }

    pub async fn run_nsql_test(
        base_url: &str,
        ts: &TestCase,
        trace_provider: &TracerProvider,
    ) -> Result<(), anyhow::Error> {
        tracing::info!("Running test cases {}", ts.name);
        let task_start_time = std::time::SystemTime::now();

        // Call /v1/nsql, check response
        let mut headers = HeaderMap::new();
        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        let response = match http_post(
            format!("{base_url}/v1/nsql").as_str(),
            &ts.body.to_string(),
            headers,
        )
        .await
        {
            Ok(response) => response,
            Err(e) => {
                tracing::error!("run_nsql_test error: {:?}", e);
                insta::assert_snapshot!(format!("{}_error", ts.name), e.to_string());
                return Ok(());
            }
        };
        insta::assert_snapshot!(format!("{}_response", ts.name), &response);

        // ensure all spans are exported into task_history
        let _ = trace_provider.force_flush();

        // Check task_history table for expected rows.
        let mut headers = HeaderMap::new();
        headers.insert(ACCEPT, HeaderValue::from_static("text/plain"));
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));

        let response = http_post(
            format!("{base_url}/v1/sql").as_str(),
            format!(
                "SELECT task, input
                        FROM runtime.task_history
                        WHERE task NOT IN ('ai_completion', 'health', 'accelerated_refresh')
                        AND start_time > '{}'
                        ORDER BY start_time, task;",
                Into::<DateTime<Utc>>::into(task_start_time).to_rfc3339()
            )
            .as_str(),
            headers,
        )
        .await
        .map_err(anyhow::Error::msg)?;

        insta::assert_snapshot!(format!("{}_tasks", ts.name), response,);

        Ok(())
    }
}

mod search {
    use http::{
        header::{ACCEPT, CONTENT_TYPE},
        HeaderMap, HeaderValue,
    };

    use crate::models::{http_post, normalize_search_response};

    pub struct TestCase {
        pub name: &'static str,
        pub body: serde_json::Value,
    }

    pub async fn run_search_test(base_url: &str, ts: &TestCase) -> Result<(), anyhow::Error> {
        tracing::info!("Running test cases {}", ts.name);

        // Call /v1/search, check response
        let mut headers = HeaderMap::new();
        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        let response_str = http_post(
            &format!("{base_url}/v1/search").to_string(),
            &ts.body.to_string(),
            headers,
        )
        .await
        .map_err(|e| anyhow::anyhow!("Failed to execute HTTP POST: {}", e))?;

        let response = serde_json::from_str(&response_str)
            .map_err(|e| anyhow::anyhow!("Failed to parse HTTP response: {}", e))?;

        insta::assert_snapshot!(
            format!("{}_response", ts.name),
            normalize_search_response(response)
        );
        Ok(())
    }
}

fn create_api_bindings_config() -> Config {
    let mut rng = rand::thread_rng();
    let http_port: u16 = rng.gen_range(50000..60000);
    let flight_port: u16 = http_port + 1;
    let otel_port: u16 = http_port + 2;
    let metrics_port: u16 = http_port + 3;

    let localhost: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));

    let api_config = Config::new()
        .with_http_bind_address(SocketAddr::new(localhost, http_port))
        .with_flight_bind_address(SocketAddr::new(localhost, flight_port))
        .with_open_telemetry_bind_address(SocketAddr::new(localhost, otel_port));

    tracing::debug!(
        "Created api bindings configuration: http: {http_port}, flight: {flight_port}, otel: {otel_port}, metrics: {metrics_port}"
    );

    api_config
}

fn get_taxi_trips_dataset() -> Dataset {
    let mut dataset = Dataset::new("s3://spiceai-demo-datasets/taxi_trips/2024/", "taxi_trips");
    dataset.params = Some(Params::from_string_map(
        vec![
            ("file_format".to_string(), "parquet".to_string()),
            ("client_timeout".to_string(), "120s".to_string()),
        ]
        .into_iter()
        .collect(),
    ));
    dataset.acceleration = Some(Acceleration {
        enabled: true,
        refresh_sql: Some("SELECT * FROM taxi_trips LIMIT 10".to_string()),
        ..Default::default()
    });
    dataset
}

fn get_tpcds_dataset(
    ds_name: &str,
    spice_name: Option<&str>,
    refresh_sql: Option<&str>,
) -> Dataset {
    let mut dataset = Dataset::new(
        format!("s3://spiceai-public-datasets/tpcds/{ds_name}/"),
        spice_name.unwrap_or(ds_name),
    );
    dataset.params = Some(Params::from_string_map(
        vec![
            ("file_format".to_string(), "parquet".to_string()),
            ("client_timeout".to_string(), "120s".to_string()),
        ]
        .into_iter()
        .collect(),
    ));
    dataset.acceleration = Some(Acceleration {
        enabled: true,
        refresh_sql: Some(
            refresh_sql
                .unwrap_or(&format!(
                    "SELECT * FROM {} LIMIT 20",
                    spice_name.unwrap_or(ds_name)
                ))
                .to_string(),
        ),
        ..Default::default()
    });
    dataset
}

/// Normalizes vector similarity search response for consistent snapshot testing by replacing dynamic
/// values such as duration with placeholder.
fn normalize_search_response(mut json: Value) -> String {
    if let Some(matches) = json.get_mut("matches").and_then(|m| m.as_array_mut()) {
        for m in matches {
            if let Some(obj) = m.as_object_mut() {
                obj.remove("score");
            }
        }
    }

    if let Some(duration) = json.get_mut("duration_ms") {
        *duration = json!("duration_ms_val");
    }

    sort_json_keys(&mut json);

    serde_json::to_string_pretty(&json).unwrap_or_default()
}

/// Normalizes embeddings response for consistent snapshot testing by replacing actual embedding arrays with a placeholder,
fn normalize_embeddings_response(mut json: Value) -> String {
    if let Some(data) = json.get_mut("data").and_then(|d| d.as_array_mut()) {
        for entry in data {
            if let Some(embedding) = entry.get_mut("embedding") {
                if let Some(embedding_array) = embedding.as_array_mut() {
                    let num_elements = embedding_array.len();
                    *embedding = json!(format!("array_{}_items", num_elements));
                } else if let Some(embedding_str) = embedding.as_str() {
                    *embedding = json!(format!("str_len_{}", embedding_str.len()));
                }
            }
        }
    }

    sort_json_keys(&mut json);

    serde_json::to_string_pretty(&json).unwrap_or_default()
}

/// Normalizes chat completion response for consistent snapshot testing by replacing dynamic values
fn normalize_chat_completion_response(mut json: Value, normalize_message_content: bool) -> String {
    // Replace `content`
    if normalize_message_content {
        if let Some(choices) = json.get_mut("choices").and_then(|c| c.as_array_mut()) {
            for choice in choices {
                if let Some(message) = choice.get_mut("message") {
                    if let Some(content) = message.get_mut("content") {
                        *content = json!("content_val");
                    }
                }
            }
        }
    }

    if let Some(created) = json.get_mut("created") {
        *created = json!("created_val");
    }

    // Replace `completion_tokens`, `prompt_tokens`, and `total_tokens` fields in `usage`
    if let Some(usage) = json.get_mut("usage") {
        if let Some(completion_tokens) = usage.get_mut("completion_tokens") {
            *completion_tokens = json!("completion_tokens_val");
        }
        if let Some(completion_tokens_details) = usage.get_mut("completion_tokens_details") {
            *completion_tokens_details = json!("completion_tokens_details_val");
        }
        if let Some(prompt_tokens_details) = usage.get_mut("prompt_tokens_details") {
            *prompt_tokens_details = json!("prompt_tokens_details_val");
        }
        if let Some(prompt_tokens) = usage.get_mut("prompt_tokens") {
            *prompt_tokens = json!("prompt_tokens_val");
        }
        if let Some(total_tokens) = usage.get_mut("total_tokens") {
            *total_tokens = json!("total_tokens_val");
        }
    }

    if let Some(system_fingerprint) = json.get_mut("system_fingerprint") {
        *system_fingerprint = json!("system_fingerprint_val");
    }

    if let Some(id) = json.get_mut("id") {
        *id = json!("id_val");
    }

    sort_json_keys(&mut json);

    serde_json::to_string_pretty(&json).unwrap_or_default()
}

/// Sorts the keys of a JSON object in place for consistent snapshot testing
fn sort_json_keys(value: &mut Value) {
    match value {
        Value::Object(map) => {
            let mut sorted_map = serde_json::Map::new();
            let mut keys: Vec<_> = map.keys().cloned().collect();
            keys.sort();

            for key in keys {
                if let Some(mut val) = map.remove(&key) {
                    sort_json_keys(&mut val); // Recurse into nested objects
                    sorted_map.insert(key, val);
                }
            }

            *map = sorted_map;
        }
        Value::Array(array) => {
            for element in array.iter_mut() {
                sort_json_keys(element);
            }
        }
        _ => {}
    }
}

async fn send_embeddings_request(
    base_url: &str,
    model: &str,
    input: EmbeddingInput,
    // The format to return the embeddings in. Can be either `float` or [`base64`](https://pypi.org/project/pybase64/). Defaults to float
    encoding_format: Option<&str>,
    // OpenAI only: A unique identifier representing your end-user, [Learn more](https://platform.openai.com/docs/usage-policies/end-user-ids).
    user: Option<&str>,
    // The number of dimensions the resulting output embeddings should have. Only supported in `text-embedding-3` and later models.
    dimensions: Option<u32>,
) -> Result<Value, reqwest::Error> {
    let mut request_body = json!({
        "model": model,
        "input": input,
    });

    if let Some(ef) = encoding_format {
        request_body["encoding_format"] = json!(ef);
    }

    if let Some(u) = user {
        request_body["user"] = Value::String(u.to_string());
    }

    if let Some(d) = dimensions {
        request_body["dimensions"] = json!(d);
    }

    let response = Client::new()
        .post(format!("{base_url}/v1/embeddings"))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await?
        .error_for_status()?
        .json::<Value>()
        .await?;

    Ok(response)
}

async fn send_chat_completions_request(
    base_url: &str,
    messages: Vec<(String, String)>,
    model: &str,
    stream: bool,
) -> Result<Value, reqwest::Error> {
    let request_body = json!({
        "messages": messages.iter().map(|(role, content)| {
            json!({
                "role": role,
                "content": content,
            })
        }).collect::<Vec<_>>(),
        "model": model,
        "stream": stream,
    });

    let response = Client::new()
        .post(format!("{base_url}/v1/chat/completions"))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await?
        .error_for_status()?
        .json::<Value>()
        .await?;

    Ok(response)
}

/// Generic function to send a POST request, returning the response as a String.
pub async fn http_post(url: &str, body: &str, headers: HeaderMap) -> Result<String, anyhow::Error> {
    let response = Client::new()
        .post(url)
        .headers(headers)
        .body(body.to_string())
        .send()
        .await
        .map_err(|e| anyhow::anyhow!("Request error: {e}"))?;

    if !response.status().is_success() {
        let status = response.status();
        let message = response
            .text()
            .await
            .unwrap_or_else(|_| "No error message".to_string());
        return Err(anyhow::anyhow!("HTTP error: {status} - {message}"));
    }

    response
        .text()
        .await
        .map_err(|e| anyhow::anyhow!("Error reading response body: {e}")) // Map body read error to anyhow
}

/// Returns a human-readable representation of the SQL query result against a [`Runtime`].
async fn sql_to_display(
    rt: &Arc<Runtime>,
    query: &str,
) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
    let data = rt
        .datafusion()
        .query_builder(query)
        .build()
        .run()
        .await
        .boxed()?
        .data
        .try_collect::<Vec<_>>()
        .await
        .boxed()?;
    pretty_format_batches(&data).map(|d| format!("{d}")).boxed()
}

#[allow(clippy::expect_used)]
async fn sql_to_single_json_value(rt: &Arc<Runtime>, query: &str) -> Value {
    let data = rt
        .datafusion()
        .query_builder(query)
        .build()
        .run()
        .await
        .boxed()
        .expect("Failed to collect data")
        .data
        .try_collect::<Vec<_>>()
        .await
        .boxed()
        .expect("Failed to collect data");
    assert_eq!(data.len(), 1);
    assert_eq!(data[0].columns().len(), 1);
    serde_json::from_str(
        data[0]
            .column(0)
            .as_any()
            .downcast_ref::<StringArray>()
            .expect("column is a StringArray")
            .value(0),
    )
    .expect("value is a JSON string")
}

async fn get_params_with_secrets_value(
    params: &HashMap<String, Value>,
    rt: &Runtime,
) -> HashMap<String, SecretString> {
    let params = params
        .clone()
        .iter()
        .map(|(k, v)| {
            let k = k.clone();
            match v.as_str() {
                Some(s) => (k, s.to_string()),
                None => (k, v.to_string()),
            }
        })
        .collect::<HashMap<_, _>>();

    get_params_with_secrets(rt.secrets(), &params).await
}
