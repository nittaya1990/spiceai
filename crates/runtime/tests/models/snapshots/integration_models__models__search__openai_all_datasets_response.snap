---
source: crates/runtime/tests/models/mod.rs
expression: normalize_search_response(response)
---
{
  "duration_ms": "duration_ms_val",
  "matches": [
    {
      "dataset": "spice.public.item",
      "primary_key": {
        "i_item_sk": 19
      },
      "value": "Patient"
    },
    {
      "dataset": "spice.public.item",
      "primary_key": {
        "i_item_sk": 7
      },
      "value": "New"
    }
  ]
}
